src/stripe/stripe.service.ts

```diff
@@ -0,0 +1,115 @@
+    const dataObject = event.data.object as Stripe.Checkout.Session;
+    console.log(`order:${dataObject.metadata.orderId} payment completed`);
+    this.completeOrderInformation(dataObject);
+  }
+
+  /**
+   * Helper function of stripeWebhookHandler, generate the payment information
+   * from stripe webhook event and create new a paymentInfo document.
+   * @param data payment information from webhook event
+   */
+  async completeOrderInformation(data: Stripe.Checkout.Session) {
+    const orderId = new mongoose.Types.ObjectId(data.metadata.orderId) as unknown as ObjectId;
+    const updateOrderStatus = await this.orderService.update(orderId, { status: "completed" });
+    if (!updateOrderStatus)
+      throw new NotFoundException(`Order: ${data.metadata.orderId} does not exist.`);
+
+    const paymentInfo: CreatePaymentInfoDto = {
+      orderId: data.metadata.orderId,
+      email: data.customer_details.email,
+      name: data.customer_details.name,
+      phone: data.customer_details.phone,
+      currency: data.currency,
+      sessionId: data.id,
+      amountTotal: data.amount_total.toString(),
+      constructionAddress: {
+        city: data.shipping_details.address.city,
+        country: data.shipping_details.address.country,
+        line1: data.shipping_details.address.line1,
+        line2: data.shipping_details.address.line2,
+        state: data.shipping_details.address.state,
+        postalCode: data.shipping_details.address.postal_code,
+      },
+      billingAddress: {
+        city: data.customer_details.address.city,
+        country: data.customer_details.address.country,
+        line1: data.customer_details.address.line1,
+        line2: data.customer_details.address.line2,
+        state: data.customer_details.address.state,
+        postalCode: data.customer_details.address.postal_code,
+      },
+    };
+    const newPayment = await this.createPaymentInfo(paymentInfo);
+    if (!newPayment) throw new Error("fail to create payment information for order!");
+  }
+
+  /**
+   * Create a new payment information document.
+   * @param data payment information from webhook event
+   * @returns new paymentInfo document created
+   */
+  async createPaymentInfo(data: CreatePaymentInfoDto): Promise<PaymentInfo> {
+    const newPaymentInfo = await this.paymentInfoModel.create(data);
+    return newPaymentInfo;
+  }
+
+  /**
+   * Find a payment information document and its relative order document.
+   * @param id order Id, not paymentInfp id
+   * @returns an object contains documents of paymentInfo and order
+   */
+  async findPaymentInfoByOrderId(
+    id: ObjectId,
+  ): Promise<{ paymentInfo?: PaymentInfo; order: Order }> {
+    const order = await this.orderService.findOne(id);
+    const paymentInfo = await this.paymentInfoModel.findOne({ orderId: id }).exec();
+    if (!order) throw new NotFoundException(`order #${id} not found`);
+    return {
+      paymentInfo,
+      order,
+    };
+  }
+
```

Comment:
My suggestion is that it's better to add a payment ref in order schema, and delete "findPaymentInfoByOrderId". Instead, you can use order.findById(id).populate().exec() to get an order's payment data.

There are two advantages to doing this:
1. When you search for multiple orders, you can get the payment data corresponding to all orders through .populate(). Without this association, it cannot be done. This feature will be used on the my order page.
2. When you later want to filter orders by attributes such as payment status/construction address, you must have this association to filter through the sub-table payment.

Reply:
> My suggestion is that it's better to add a payment ref in order schema, and delete "findPaymentInfoByOrderId". Instead, you can use order.findById(id).populate().exec() to get an order's payment data.
> 
> There are two advantages to doing this:
> 
> 1. When you search for multiple orders, you can get the payment data corresponding to all orders through .populate(). Without this association, it cannot be done. This feature will be used on the my order page.
> 2. When you later want to filter orders by attributes such as payment status/construction address, you must have this association to filter through the sub-table payment.

fixed, check the findAll function in the order provider (orders/order.service.ts)